src/gt4py/next/program_processors/runners/dace_iterator/__init__.py

```diff
@@ -226,24 +235,22 @@ def run_dace_iterator(program: itir.FencilDefinition, *args, **kwargs) -> None:
 
 
 @program_executor
-def run_dace_cpu(program: itir.FencilDefinition, *args, **kwargs) -> None:
-    run_dace_iterator(
-        program,
-        *args,
-        **kwargs,
-        build_cache=_build_cache_cpu,
-        build_type=_build_type,
-        run_on_gpu=False,
-    )
+def run_dace(program: itir.FencilDefinition, *args, **kwargs) -> None:
+    run_on_gpu = any(not isinstance(arg.ndarray, np.ndarray) for arg in args if is_field(arg))
+    if run_on_gpu:
+        if cp is None:
+            raise RuntimeError(
+                f"Non-numpy field argument passed to program {program.id} but module cupy not installed"
+            )
 
+        if not all(isinstance(arg.ndarray, cp.ndarray) for arg in args if is_field(arg)):
+            raise RuntimeError("Execution on GPU requires all fields to be stored as cupy arrays")
```

Comment:
Yeah, I like this direction much better. We could do this in a nicer way, perhaps using the array API, because this won't really scale if we want to add support for JAX, tensorflow, PyTorch and the like's own arrays. Anyways, that I think is outside the scope of this PR.